src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs

```diff
@@ -61,51 +61,32 @@ public void ClearSessionScopeTask()
-                        return compilationAnalysisScope;
-                    }, analyzerExecutor.CancellationToken);
-                };
-
-                var callback = new ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>>.CreateValueCallback(getTask);
-                var compilationActionsCache = GetOrCreateCompilationActionsCache(analyzerExecutor.AnalyzerOptions);
-                return compilationActionsCache.GetValue(analyzerExecutor.Compilation, callback);
-            }
-
-            private ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>> GetOrCreateCompilationActionsCache(AnalyzerOptions analyzerOptions)
-            {
                 lock (_gate)
                 {
-                    ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>> compilationActionsCache;
-
-                    if (_lazyCompilationScopeMap == null)
+                    _lazyCompilationScopeCache = _lazyCompilationScopeCache ?? new ConditionalWeakTable<Compilation, Dictionary<AnalyzerOptions, Task<HostCompilationStartAnalysisScope>>>();
+                    var compilationActionsMap = _lazyCompilationScopeCache.GetOrCreateValue(analyzerExecutor.Compilation);
+                    Task<HostCompilationStartAnalysisScope> task;
+                    if (!compilationActionsMap.TryGetValue(analyzerExecutor.AnalyzerOptions, out task))
                     {
-                        _lazyCompilationScopeMap = new Dictionary<AnalyzerOptions, ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>>>();
-                    }
-                    else if (_lazyCompilationScopeMap.TryGetValue(analyzerOptions, out compilationActionsCache))
-                    {
-                        return compilationActionsCache;
+                        task = Task.Run(() =>
+                        {
+                            var compilationAnalysisScope = new HostCompilationStartAnalysisScope(sessionScope);
+                            analyzerExecutor.ExecuteCompilationStartActions(sessionScope.CompilationStartActions, compilationAnalysisScope);
+                            return compilationAnalysisScope;
+                        }, analyzerExecutor.CancellationToken);
+
+                        compilationActionsMap.Add(analyzerExecutor.AnalyzerOptions, task);
                     }
 
-                    compilationActionsCache = new ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>>();
-                    _lazyCompilationScopeMap.Add(analyzerOptions, compilationActionsCache);
-                    return compilationActionsCache;
+                    return task;
                 }
             }
 
-            public void ClearCompilationScopeMap(AnalyzerOptions analyzerOptions, Compilation compilation)
+            public void ClearCompilationScopeMap(Compilation compilation)
             {
                 lock (_gate)
```

Comment:
:bulb: This lock is now technically unnecessary (`ConditionalWeakTable` provides synchronization), but I think it would be confusing to some readers to remove it.